# 11. 제너레이터

## 제너레이터란

제너레이터는 iterator를 생성해주는 함수이다. \
함수 안에서 yield를 사용하면 함수는 제너레이터가 되며 yield에는 값(변수)을 지정한다.

제너레이터는 다음과 같은 특징이 있다.

* 느슨하게 평가된다.(순서의 다음 값은 제너레이터가 다음 값을 호출할 때 실행됨)
* 함수의 내부 로컬 변수를 통해 내부상태가 유지된다.
* 무한한 순서가 있는 객체를 만들 수 있다.(명확한 끝이 없는 데이터 스트림)

아래는 제너레이터 예제이다.

```python
def number_generator():
    yield 0
    yield 1
    yield 2
 
for i in number_generator():
    print(i)
    
# 출력 결과
# 0
# 1
# 2
```

## 제너레이터와 일반 함수와의차이점

일반 함수는 호출될 때, 함수에 작성된 코드가 전부 실행되고 return 문에서 반환값을 반환한다. 제너레이터는 첫 호출 시에는 함수의 처음부터 yield 문을 처음 만나기 전까지의 코드만 실행이 된다. 이 때, yield 문에 반환할 값이 설정되어 있다면 이를 반환한다. 이후, 다시 제너레이터가 호출되면 마지막으로 yield가 호출되었던 코드 다음 부터 그 다음 yield를 만날 때 까지 코드를 진행한다. 이를 반복하다 함수가 끝까지 진행되거나 return문을 만나면 더이상 값을 반환하지 않는다.

제너레이터는 내부의 모든 코드가 실행되고 종료된 후에는 더이상 데이터를 반환하지 않는다. 이후 해당 제너레이터가 값을 반환할 수 없어 호출할 경우 에러가 발생한다. 따라서 제너레이터 내부의 모든 데이터가 처리되고 난 이후에는 새로이 제너레이터를 생성해야 된다.

다음 코드는 제너레이터의 모든 코드가 실행된 이후, 해당 제너레이터를 계속 호출 했을 때 발생하는 에러 예제이다.

```python
def number_generator():
    yield 0
    yield 1
    yield 2
    
gen = number_generator()
next(gen) # 0
next(gen) # 1
next(gen) # 2
next(gen) # error StopIteration 발생
```



## yield from

yield from은 서로 다른 제너레이터를 합쳐서 사용할 수 있게 해주는 키워드다.

```python
def gen(): 
    yield from range(10) 
    yield from range(5)
```

위의 예시와 같이 각각 다른 제너레이터를 한 함수에 같이 선언하여 사용할 수 있다. 제너레이터를 합칠때에는 무한한 제너레이터와 유한한 제너레이터를 합칠때 주의해야 한다. 만약 첫 호출 순서로 무한 제너레이터가 있다면 두번째 제너레이터가 값을 반환할 차례는 오지 않는다.&#x20;

예를 들어서 첫번째 제너레이터는 itertools의 count메소드를 이용해 초기값과 증가값을 가진 상태라고 가정하고 두번째 제너레이터는 1부터 10까지의 각 수의 제곱을 반환해주는 제너레이터가 있다고 가정해본다. 이 때, 첫 제너레이터는  무한한 값을 반환하는 제너레이터이기 때문에 아무리 호출을 해도 두번째 제너레이터는 값을 반환할 수 없게 된다
